### ISSUE TITLE
Health

### ISSUE THREAD

**Issue by kodomazer** [evolvio/evolv.io#59]
In order to promote fighting I propose we add in a separate counter for the "health" of a creature.  When health reaches 0 the creature dies and all of its yums fall to the ground.

I believe that the reason why there are no creatures that evolve to kill is that it is not economical to do so.  The only time you eat is when you've killed, it costs you to attack and you don't get the kill until after all of the energy is destroyed.

**Comment by PRabahy** [comment]
I like the idea of having health, but one of your assumptions is untrue. A creature doesn't have to completely kill another creature (in the current code) to get a benefit. If you look at Creature.fight, the attacking creature looses [fightLevel * FIGHT_ENERGY * energy * timeStep] energy and the attacked creature(s) drop [fightLevel * INJURED_ENERGY * timeStep] energy.

fightLevel and timeStep are in both equations, so they cancel each other out.
FIGHT_ENERGY = 0.06;
INJURED_ENERGY = 0.25;
energy = the size of the attacking creature at the time.

So basically, it never makes sense to attack if you are 4x bigger than the creatures you are attacking, but the current vision system doesn't return the energy level of creatures that you can see. Also, if you ever attack when there are no other creatures around, you still loose all the FIGHT_ENERGY, but don't get any benefit at all.

**Comment by PRabahy** [comment]
Oops, I also left out that even after a creature successfully attacks another causing it to drop energy, it still needs to eat the energy. That ends up having a penalty of [double EAT_ENERGY = 0.05] (assuming a perfect match to mouth hue). In the end it takes 0.11 energy to have the potential to gain of 0.25 (best case) by fighting.

**Comment by kodomazer** [comment]
Ah, I didn't read the code completely at the time of posting.  I assumed the energy just got lost.  In light of this we have a few options to try to make fighting more lucrative:

- Reduce cost of fighting, possibly making it a flat cost instead of a fraction of total energy
- Implement health and possibly take out the dropping of energy until the very end
- Do nothing/Put this on the back burner for now

**Comment by PRabahy** [comment]
I think option 2 makes the most sense to me. Unfortunately, I believe that change would make it very unlikely to evolve a creature that has a fighting disposition. If we did that, I would want to make sure that we also enhance vision so that it can see the energy and health of creatures in its line of site.

https://github.com/evolvio/evolv.io/pull/46 was my first attempt to improve fighting. Basically, if a creature's carnivore ratio was higher, it would be more beneficial for it to fight.

**Comment by kodomazer** [comment]
Something that we could do is leave the bodies lying around and if you "fight" with a corpse in range then you end up eating a chunk of it, maybe a tenth of your current energy or however much energy is left on the corpse.  With corpses just give them a linear "rotting" rate as well, so we don't end up with too many bodies lying around.

**Comment by pilger15** [comment]
Another problem is, that the injured creature can eat its own energy. 
Furthermore the energy lost fighting depends on its total energy? This is counter-intuitive and makes it impossible for fighters to get big.

I would suggest increasing cost and benefit of fighting.
e.g:
- FIGHT_ENERGY = EAT_ENERGY * 5;
- INJURED_ENERGY = FIGHT_ENERGY * 3;


and changing the lost energy to in the fighting function to
      loseEnergy(fightLevel * FIGHT_ENERGY * timeStep); 
and make the injured creature lose Energy instead of dropping it and adding it straight to the attacker:
   double injuredEnergy = fightLevel * INJURED_ENERGY * timeStep;
    ((Creature)collider).loseEnergy(injuredEnergy);
    addEnergy(injuredEnergy); 

However, for this to work, it would be necessary to add the occlusion distances of the views as inputs, so attackers can decide when attacking is sensible (occlusion distance very small) and when it would be just a waste of energy.

Lastly: 
If the injured creature dies the 100Energy could be added to the attacker.
However, I tested that and in very long term it resulted in a creature that grew dramatically and instantly ate all creatures. So a maximum size or something is needed.

**Comment by UnknownEvil** [comment]
I know I'm a bit late here, but you can get some creatures to attack on occasion if you just remove the timeStep *= 100 on line 44 in CreatureAction

**Comment by UnknownEvil** [comment]
I don't believe the creatures get a value based on if another creature is in their path, perhaps each eye should get one, and thus allowing creatures to form a sort of line follower, to follow creatures
